### ISSUE TITLE
prevent flapping

### ISSUE THREAD

**Issue by pgte** [libp2p/js-libp2p-connection-manager#8]
In some circumstances, these events might happen in a lopp:

* a measure is slightly above a threshold
* that triggers a preemptive disconnect
* the measure goes down to slightly below the threshold
* a new connection is accepted
* repeat

To prevent this,  dampening should be reflected on the measure. In some cases, it already is.

By default, some statistics (like data sent and received) uses the moving average over one minute, which already produces a dampening effect.

The problem is that the other statistics are a snapshot taken at a given point in time, like the number of connected peers. In my opinion, this value, like the moving averages, should be continuous. Let me illustrate:

This peer is connected to 3 other peers. A new peer connection happens. The internal number of connected peers does not suddenly become 4. It gradually raises to 4, over a continuum.

This should prevent flapping.

**Comment by serzero2007** [comment]
Exactly the same problem appears in devices like refrigerator temperature controller, where it is solved by using two different temperatures to stop cooling and start again. Basically, https://en.m.wikipedia.org/wiki/Bang–bang_control

In short we have to have two threshold values with a considerable gap between them --- one for the purpose of forcing disconnecting and the other --- for the purpose of allowing peers to connect again.

**Comment by pgte** [comment]
@serzero2007 thank you for showing me this. So you're saying that we need two values: one high value that triggers the "cooling down" phase and another (lower value) below which we can stop the cooling and accept new connections again. That sounds good!

I think that the measure should be continuous (integrated over time), in the sense that, for instance if the value of connected peers raises from 3 to 4, the actual measured value raises from 3 to 4 in a continuous way (using exponential moving average, for instance).

This way, the user of the system could define a target maximum number of peers and, instead of using that one hard limit (which would cause flapping), we could somehow come up with the two values with a gap between, without being limited to integer numbers.

Makes sense?

**Comment by serzero2007** [comment]
There is a problem with using this kernel function: exp(-t/tau). It only affects the time when the state transition happens a little: on the time axis effect won't be more than tau.

But at the other hand, working with floats and calculating F_smooth all the time seems much slower than working with integers.

**Comment by serzero2007** [comment]
A, B, C to be estimated so the system has desired properties. But anyway, that's kind of overkill for this task. It looks like the best option is the following:
Let  X ∈ N = [ 0, 1, 2,  ... ]   --- Measure.
       X_1 = 8,  --- the lowest of two values
       X_2 = 14 --- the highest
       T  = 200  --- Throttling time (in milliseconds) 
(Obviously we can only limit the pace of establishing new connections)
What's the most important here --- we do not have to do heavy calculations. 
At the same time we have very smooth process of establishing connections 
from 0 to X_2. thanks to throttling.

What also seems important to me, but unrelated to this topic, is that first of all we need to distinguish useful peers from not so useful and treat them differently. At least:
1) The peer we communicated to directly must have higher weight than the peer we have never dialed.
2) the peer that knows more peers > the peer that knows less peers
3) the peers that support protocols we support > the peer that don't
Reliability of the connection depends on this "subjective quality of peers" a lot.

**Comment by pgte** [comment]
@serzero2007 thank you for these insights, they're going to be really useful when choosing an implementation!

About the peer value: you're right, and these concerns are being taken care of in other issues.
For instance, for a peer that is relaying to other peers, we have [this issue](https://github.com/libp2p/js-libp2p-connection-manager/issues/10). Otherwise, our current approach is to let the application define the peer value for peers we're connected to. This would inform the connection manager of which connections to remove first if needed. Makes sense?

**Comment by serzero2007** [comment]
I have just realized the following idea:
1) Define separate peer value function for each subsystem of libp2p.
2) Define independent peer limits for each subsystem of libp2p.
3) Unless peer has value at least to 1 subsystem, it cannot be kicked.
4) [Optional] Define rules to force kicking any peer that satisfy them
So we need something like the following API:
```ts
enum RuleType {
  'PreservingRule',
  'BlacklistingRule'
}

type Rule = [ RuleType, Function, { enabled?: boolean, limit?: number, ... }]

class ConnectionManager {
  private rules: Map<tag, Rule> = new Map();

  // Adding Rules to Rule-Set
  addPreservingRule (tag: String, fn: (peer:PeerId) => number, limit: number): void;
  addBlacklistingRule (tag: String, fn: (peer:PeerId, peerValues: Object) => boolean): void;

  // Modifing Rules. Disabeling/Enabling them 
  setLimit(tag: String, limit: number): void;
  deleteRule(tag: String): void;
  enableRule(tag: String): void;
  disableRule(tag: String): void;

  // Appling rules to peers.
  // if tag == undefined then all rulles are applied  
  checkPeer(peer, tag?: String): boolean;

  // Automatic checking of peers
  // throttleInterval == checkInterval by default
  async start(checkInterval: number, throttleInterval?: number): Promise<boolean>;
  async stop(): Promise<boolean>;
}
```

**Comment by serzero2007** [comment]
if we allow `tag` to be `String[]`, we can have namespacing different parts of libp2p by means of that:
```ts
connectionManager.addPreservingRule(
  [ 'DHT', 'PRESERVE_IMPORTANT_FOR_DHT' ], function (...) {}, 10
);
connectionManager.addPreservingRule(
  [ 'DHT', 'PRESERVE_SERVER_PEERS' ], function (...) {}, 10
);
disableRule([ 'DHT']) // Disables both rules
enableRule([ 'DHT', 'PRESERVE_SERVER_PEERS' ]) // Enables a specific rule
```

**Comment by serzero2007** [comment]
The main idea here is that the only situation when we need some peers not to be disconnected, is a situation when we have to preserve them for some specific reason.

Different reasons manifest themselves as different rules in this implementation proposal.

**Comment by pgte** [comment]
@serzero2007 That's true, right now we're delegating entirely on the application to know and preserve some nodes that may be important to keep some subsystems working (like DHT).

I've created a separate issue for this here: https://github.com/libp2p/js-libp2p-connection-manager/issues/12. Could you please add your ideas there?

**Comment by Kubuxu** [comment]
@pgte please see how go-libp2p does it: https://github.com/libp2p/go-libp2p-connmgr/blob/master/connmgr.go#L61

We have following parameters low-water, high-water and grace period. We trigger connection closing when we reach high-water, we close connections until we have low-water connections but we don't close connections that live shorter than grace period.

See also: https://github.com/ipfs/go-ipfs/blob/master/docs/config.md#connmgr
